third_party/move/move-vm/types/src/utils/union_find.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#[derive(Debug)]
+pub struct UnionFind {
+    parent: Vec<u32>,
+    rank: Vec<u8>,
+    num_sets: usize,
+    capacity: usize,
+}
+
+impl UnionFind {
+    /// Creates a new UnionFind instance with `num_elements` and `capacity`
+    pub fn new_with_size_and_capacity(
+        num_elements: usize,
+        capacity: usize,
+    ) -> Result<Self, &'static str> {
+        if num_elements > std::u32::MAX as usize {
+            return Err("num_elements larger than std::u32::MAX");
+        }
+        if capacity > std::u32::MAX as usize {
+            return Err("capacity larger than std::u32::MAX");
+        }
+        if num_elements > capacity {
+            return Err("num_elements larger than given capacity");
```

Comment:
Again, we can just say "The number of elements" without referring to variable name?

Reply:
also we can embed the actual values in the error string (with format!)

Reply:
if we check this first, we only need to check capacity < MAX right?